plugins/renepay/pay.c

```diff
@@ -0,0 +1,1798 @@
+		 *     1970-01-01 UTC is greater than `created_at` plus
+		 *     `seconds_from_creation`.
+		 * - otherwise:
+		 *   - MUST reject the invoice if the current time since
+		 *     1970-01-01 UTC is greater than `created_at` plus
+		 * 7200.
+		 */
+		if (b12->invoice_relative_expiry)
+			invexpiry = *b12->invoice_created_at + *b12->invoice_relative_expiry;
+		else
+			invexpiry = *b12->invoice_created_at + BOLT12_DEFAULT_REL_EXPIRY;
+	}
+
+	if (node_id_eq(&pay_plugin->my_id, &p->destination))
+		return renepay_fail(renepay, JSONRPC2_INVALID_PARAMS,
+				    "This payment is destined for ourselves. "
+				    "Self-payments are not supported");
+
+
+	// set the payment amount
+	if (invmsat) {
+		// amount is written in the invoice
+		if (msat) {
+			return renepay_fail(renepay, JSONRPC2_INVALID_PARAMS,
+					    "amount_msat parameter unnecessary");
+		}
+		p->amount = *invmsat;
+		tal_free(invmsat);
+	} else {
+		// amount is not written in the invoice
+		if (!msat) {
+			return renepay_fail(renepay, JSONRPC2_INVALID_PARAMS,
+					    "amount_msat parameter required");
+		}
+		p->amount = *msat;
+		tal_free(msat);
+	}
+
+	/* Default max fee is 5 sats, or 0.5%, whichever is *higher* */
+	if (!maxfee) {
+		struct amount_msat fee = amount_msat_div(p->amount, 200);
+		if (amount_msat_less(fee, AMOUNT_MSAT(5000)))
+			fee = AMOUNT_MSAT(5000);
+		maxfee = tal_dup(tmpctx, struct amount_msat, &fee);
+	}
+
+	if (!amount_msat_add(&p->maxspend, p->amount, *maxfee)) {
+		return renepay_fail(
+			renepay, JSONRPC2_INVALID_PARAMS,
+			"Overflow when computing fee budget, fee far too high.");
+	}
+	tal_free(maxfee);
+
+	const u64 now_sec = time_now().ts.tv_sec;
+	if (now_sec > invexpiry)
+		return renepay_fail(renepay, PAY_INVOICE_EXPIRED, "Invoice expired");
+
+	/* To construct the uncertainty network we need to perform the following
+	 * steps:
+	 * 1. check that there is a 1-to-1 map between channels in gossmap
+	 * and the uncertainty network. We call `uncertainty_network_update`
+	 *
+	 * 2. add my local channels that could be private.
+	 * We call `update_uncertainty_network_from_listpeerchannels`.
+	 *
+	 * 3. add hidden/private channels listed in the routehints.
+	 * We call `uncertainty_network_add_routehints`.
+	 *
+	 * 4. check the uncertainty network invariants.
+	 * */
+	if(gossmap_changed)
+		uncertainty_network_update(pay_plugin->gossmap,
+					   pay_plugin->chan_extra_map);
+
+
+	/* TODO(eduardo): We use a linear function to decide how to decay the
+	 * channel information. Other shapes could be used.
+	 * Also the choice of the proportional parameter TIMER_FORGET_SEC is
+	 * arbitrary.
+	 * Another idea is to measure time in blockheight. */
+	const double fraction = (now_sec - pay_plugin->last_time)*1.0/TIMER_FORGET_SEC;
+	uncertainty_network_relax_fraction(pay_plugin->chan_extra_map,
+					   fraction);
+	pay_plugin->last_time = now_sec;
+
+	// TODO(eduardo): are there route hints for B12?
```

Comment:
No.  We have blinded paths instead, which *must* be used.  Though, a node can just include a dummy blinded path to/from itself.

So, if it doesn't, we need to isolate the destination from any real channels, and create fake "channels" from the entry point using the blinded paths (they have info on fees and cltv).  But that's just typing, basically, since algo stays the same.

Reply:
can we have several blinded paths which do not sum to the full but a higher amount? basically like receivers MPP split? In that case the minimum cost flow problem could be redefined as a single source multi destination problem but the constraints become weired as the blinded paths probably expect either the full amount or no amount at all.

Reply:
Yep, you can offer multiple paths with lesser capacity, it will Just Work.  Or multiple paths each with sufficient capacity, and sender chooses how to use them.